### ISSUE TITLE
Serialize an enum as an object

### ISSUE THREAD

**Issue by joffrey-bion** [Kotlin/kotlinx.serialization#1559]
**What is your use-case and why do you need this feature?**

I have an enum whose JSON representation should be a JSON object with the enum's properties instead of a JSON string representing the enum constant's name:

```kotlin
enum class GriffonIndicator(
        val level: String,
        val backgroundColor: String,
        val textColor: String,
) {
    ND(level = "ND", backgroundColor = "#ffffff", "#000000"),
    WEAK(level = "Faible", backgroundColor = "#2196F3", "#ffffff"),
    LIGHT(level = "Léger", backgroundColor = "#4CAF50", "#ffffff");
}
```

I want to serialize `WEAK` as:

```
{
   "level": "Faible",
   "backgroundColor": "#2196F3",
   "textColor": "#ffffff"
}
```

Is there already a way to do this without a custom serializer?

**Describe the solution you'd like**

I'm looking for something similar to Jackson's `@JsonFormat(shape = JsonFormat.Shape.OBJECT)`.

**Comment by sandwwraith** [comment]
No, we do not have nor plan to have a structure-like enum serializer (see also https://github.com/Kotlin/kotlinx.serialization/issues/1398). Custom serializer is a way to go

**Comment by joffrey-bion** [comment]
Thanks for your reply! Could you please explain the reason why it's not in the plans at all?

I believe it's not an uncommon use case when you have a regular class from the point of view of the API consumer, but it's used as an enum on the server side because there is a well-known number of instances.
If both sides use kotlin and share the data model, this is not a problem of course, but if you're building an API consumed by external parties, you might not want to expose that this structure is an enum on your side (it may be an implementation detail for you, and you may plan to later turn this enum into a data class and load instances from a database instead).

I considered the option of a data class + a bunch of constants, but then I lose the ability to "find" an enum value matching a predicate. I would have to add all instances to a list that I maintain myself, which is brittle because I could declare one constant and forget to update the list.

> Custom serializer is a way to go

Do you mean manually defining the serial descriptor for the class etc? It might be ok for people who want to serialize the enum as one of its properties, but I would definitely never use a custom serializer for that in my case, as it's way too complicated for such a simple use case. At this point it becomes easier to just copy-paste the enum class and create a twin data class + a conversion method, and convert enum instances into the data class instances before serializing (which is a bad approach, but simpler to write, which goes against "the simple way should be the correct one").

Or maybe I missed something and there are facilities that help building such a serializer in a simpler way?
